### PR TITLE
Introduce autotuning to `conv2d` and `conv_transpose2d` with a new `im2col`/`GEMM` algorithm

### PR DESCRIPTION
**Pull request by wingertge**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

This goes some way to resolving this old enhancement suggestion: #805 

### Changes

Adds the required infrastructure to autotune `conv2d` and `conv_transpose2d`, as well as adding a second algorithm based on `im2col` which provides significant speedups at the cost of memory usage. I'd like to add more algorithms when I have time, but this will already put in the infrastructure to make that much easier and less breaking.

#### Update; Now includes implicit GEMM when:
- CMMA is available
- `batch_size * out_h * out_w` is divisible by 16
- `out_channels` is divisible by 16
- `in_channels * kernel_h * kernel_w` is divisible by 16

### Testing
The autotuned, direct (current) and `im2col` implementations pass all existing tests (except matches_reference_backend, see below).

#### Direct
`batch_size = 4`
| Benchmark | Feature     | Backend             | Device                  | Median         |
|-----------|-------------|---------------------|-------------------------|----------------|
| conv2d    | wgpu        | `jit<wgpu>`         | BestAvailable           | 8.688ms        |
| conv2d    | wgpu-fusion | `fusion<jit<wgpu>>` | BestAvailable           | 8.351ms        |
| conv2d    | cuda-jit    | `jit<cuda>`         | CudaDevice { index: 0 } | 6.250ms        |

`batch_size = 16`
| Benchmark | Feature     | Backend             | Device                  | Median         |
|-----------|-------------|---------------------|-------------------------|----------------|
| conv2d    | wgpu        | `jit<wgpu>`         | BestAvailable           | 31.056ms       |
| conv2d    | wgpu-fusion | `fusion<jit<wgpu>>` | BestAvailable           | 32.943ms       |
| conv2d    | cuda-jit    | `jit<cuda>`         | CudaDevice { index: 0 } | 26.165ms       |

#### Im2col
`batch_size = 4`
| Benchmark | Feature     | Backend             | Device                  | Median         |
|-----------|-------------|---------------------|-------------------------|----------------|
| conv2d    | wgpu        | `jit<wgpu>`         | BestAvailable           | 4.831ms        |
| conv2d    | wgpu-fusion | `fusion<jit<wgpu>>` | BestAvailable           | 4.759ms        |
| conv2d    | cuda-jit    | `jit<cuda>`         | CudaDevice { index: 0 } | 2.617ms        |

`batch_size = 16 (split into 4 sub-batches)`
| Benchmark | Feature     | Backend             | Device                  | Median         |
|-----------|-------------|---------------------|-------------------------|----------------|
| conv2d    | wgpu        | `jit<wgpu>`         | BestAvailable           | 21.284ms       |
| conv2d    | wgpu-fusion | `fusion<jit<wgpu>>` | BestAvailable           | 21.207ms       |
| conv2d    | cuda-jit    | `jit<cuda>`         | CudaDevice { index: 0 } | 17.202ms       |

#### Implicit GEMM
`batch_size = 16`
| Benchmark | Feature  | Backend     | Device                  | Median         |
|-----------|----------|-------------|-------------------------|----------------|
| conv2d    | cuda-jit | `jit<cuda>` | CudaDevice { index: 0 } | 4.859ms        |

### CODE REVIEW

**Review by nathanielsimard — COMMENTED** [review]
LGTM, waiting for @louisfd to review before merging, but thanks a lot 🙏

File: crates/burn-jit/src/kernel/conv/conv2d/direct.rs
Comment (louisfd) [inline crates/burn-jit/src/kernel/conv/conv2d/direct.rs:52]: Is this the fix for the bug you mentioned about not having the exact same result as torch?

File: crates/burn-jit/src/kernel/conv/conv2d/im2col.rs
Comment (louisfd) [inline crates/burn-jit/src/kernel/conv/conv2d/im2col.rs:75]: Are there configurations where we are guaranteed to always be within bounds? Is so we could have a comptime strategy to remove the if when not necessary.

File: crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs
Comment (louisfd) [inline crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs:20]: I think it's fine to hardcode these values for now, but they should be passed to the cube kernel via a comptime struct, because the day they will change (for amd it's 64 for instance) we won't have to change the kernels, just the launch. Same for other consts below.

File: crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs
Comment (louisfd) [inline crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs:31]: I'm genuinely curious, why did you choose (128, 2) and not (32, 8) for instance, is it better?

File: crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs
Comment (louisfd) [inline crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs:157]: Doesn't arithmetic on two comptime values always give a comptime? Why the need for a comptime macro?

File: crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs
Comment (louisfd) [inline crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs:256]: We will probably want a comptime decision to whether or not have to do this if.

File: crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs
Comment (louisfd) [inline crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs:291]: We will probably want a comptime decision to whether or not have to do this if.

File: crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs
Comment (louisfd) [inline crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs:292]: When doing the matmul, i found you can safely reuse a fragment for several executions. So this line could go in the outer loop.

File: crates/burn-jit/src/kernel/conv/conv2d/col2im.rs
Comment (louisfd) [inline crates/burn-jit/src/kernel/conv/conv2d/col2im.rs:134]: Why not start the convolution with bias as the input?

**Review by louisfd — CHANGES_REQUESTED** [review]
I have a few questions, but overall it's very good, awesome!

File: crates/burn-jit/src/kernel/conv/conv2d/base.rs
Comment (wingertge) [reply crates/burn-jit/src/kernel/conv/conv2d/base.rs:20]: Fun fact: that typo is also on `MatmulStrategy`, which is how it got into my code (I copy pasted). I'll fix it in both.

File: crates/burn-jit/src/kernel/conv/conv2d/direct.rs
Comment (wingertge) [reply crates/burn-jit/src/kernel/conv/conv2d/direct.rs:52]: Exactly.

File: crates/burn-jit/src/kernel/conv/conv2d/im2col.rs
Comment (wingertge) [reply crates/burn-jit/src/kernel/conv/conv2d/im2col.rs:75]: Since we're iterating over out dimensions it should always be in bounds as long as padding == 0. I'll add a comptime path for that.

File: crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs
Comment (wingertge) [reply crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs:20]: Yeah I wanted to read these constants from `cubecl` ideally. I guess I'll just put them in a struct for now then so it's easier to change later.

File: crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs
Comment (wingertge) [reply crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs:157]: That was left over from a refactor, I'll just remove it.

File: crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs
Comment (wingertge) [reply crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs:31]: It makes sense when you look at the dimensionality of the matrix. The `x` dimension is `batch_size * out_h * out_w` which is much larger than the `y` dimension, which is `in_channels * kernel_h * kernel_w`. `128 x 2` is a decent approximation of the relative sizes between `M` and `N`. This lopsidedness caused issues before with the `im2col` because the matmul kernel would pick something like `144k x 1` as its cube count because it uses a square `CubeDim` but the actual matrix is anything but square.

File: crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs
Comment (wingertge) [reply crates/burn-jit/src/kernel/conv/conv2d/implicit_gemm.rs]: I'm not sure what you mean? I am reusing the fragments, but I still need to load a tile and execute the matmul for each `k`.

File: crates/burn-jit/src/kernel/conv/conv2d/col2im.rs
Comment (wingertge) [reply crates/burn-jit/src/kernel/conv/conv2d/col2im.rs:134]: Because this is faster. The columns matrix contains a lot of duplicates, so you would end up adding the bias several times for each pixel. Adding it at the end takes far fewer operations.

**Review by louisfd — APPROVED** [review]
LGTM

**Pull request merged**
